sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveSchemaInferenceSuite.scala

```diff
@@ -0,0 +1,162 @@
+
+  // Creates a case-sensitive external Hive table for testing schema inference options. Table
+  // will not have Spark-specific table properties set.
+  private def setupCaseSensitiveTable(
+      tableName: String,
+      dir: File): Unit = {
+    spark.range(NUM_RECORDS)
+      .selectExpr("id as fieldOne", "id as partCol1", "id as partCol2")
+      .write
+      .partitionBy("partCol1", "partCol2")
+      .mode("overwrite")
+      .parquet(dir.getAbsolutePath)
+
+    val lowercaseSchema = StructType(Seq(
+      StructField("fieldone", LongType),
+      StructField("partcol1", IntegerType),
+      StructField("partcol2", IntegerType)))
+
+    val client = spark.sharedState.externalCatalog.asInstanceOf[HiveExternalCatalog].client
+
+    val catalogTable = hiveExternalCatalogTable(
+      tableName,
+      dir.getAbsolutePath,
+      lowercaseSchema,
+      Seq("partcol1", "partcol2"))
+    client.createTable(catalogTable, true)
+
+    val partitions = (0 until NUM_RECORDS).map(hiveCatalogPartition(dir.getAbsolutePath, _)).toSeq
+    client.createPartitions("default", tableName, partitions, true)
+  }
+
+  // Create a test table used for a single unit test, with data stored in the specified directory.
+  private def withTestTable(dir: File)(f: File => Unit): Unit = {
+    setupCaseSensitiveTable(TEST_TABLE_NAME, dir)
+    try f(dir) finally spark.sql(s"DROP TABLE IF EXISTS $TEST_TABLE_NAME")
+  }
+
+  override def beforeEach(): Unit = {
+    super.beforeEach()
+    FileStatusCache.resetForTesting()
+  }
+
+  override def afterEach(): Unit = {
+    super.afterEach()
+    FileStatusCache.resetForTesting()
+  }
+
+  test("Queries against case-sensitive tables with no schema in table properties should work " +
+    "when schema inference is enabled") {
+    withSQLConf("spark.sql.hive.schemaInferenceMode" -> "INFER_AND_SAVE") {
+      withTempDir { dir =>
+        withTestTable(dir) { dir =>
+          val expectedSchema = StructType(Seq(
+            StructField("fieldOne", LongType),
+            // Partition columns remain case-insensitive
+            StructField("partcol1", IntegerType),
+            StructField("partcol2", IntegerType)))
+          assert(spark.sql(FIELD_QUERY).count == NUM_RECORDS)
+          assert(spark.sql(PARTITION_COLUMN_QUERY).count == NUM_RECORDS)
+          // Test that the case-sensitive schema was storied as a table property after inference
+          assert(spark.sql(SELECT_ALL_QUERY).schema == expectedSchema)
+        }
+      }
+    }
+  }
+
+  test("Schema should be inferred but not stored when ...") {
+    withSQLConf("spark.sql.hive.schemaInferenceMode" -> "INFER_ONLY") {
+      withTempDir { dir =>
+        withTestTable(dir) { dir =>
+          val existingSchema = spark.sql(SELECT_ALL_QUERY).schema
+          assert(spark.sql(FIELD_QUERY).count == NUM_RECORDS)
+          assert(spark.sql(PARTITION_COLUMN_QUERY).count == NUM_RECORDS)
+          assert(spark.sql(SELECT_ALL_QUERY).schema == existingSchema)
```

Comment:
We should check if catalog table is not updated when `schemaInferenceMode` is `INFER_ONLY`.